File: apps/api/app/routes/notifications.py
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.dependencies import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.notification import (
    MarkAllReadResponse,
    NotificationListResponse,
    NotificationRead,
    UnreadCountResponse,
)
from app.services import notifications_service as svc

router = APIRouter(prefix="/api/notifications", tags=["notifications"])


@router.get("", response_model=NotificationListResponse)
async def list_notifications(
    unread_only: bool = False,
    limit: int = Query(default=20, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> NotificationListResponse:
    return await svc.list_notifications(
        session, user.id, unread_only=unread_only, limit=limit, offset=offset
    )


@router.get("/unread-count", response_model=UnreadCountResponse)
async def unread_count(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> UnreadCountResponse:
    return await svc.get_unread_count(session, user.id)


@router.patch("/{notification_id}/read", response_model=NotificationRead)
async def mark_read(
    notification_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> NotificationRead:
    n = await svc.mark_read(session, user.id, notification_id)
    if n is None:
        raise HTTPException(status_code=404, detail="notification not found")
    return n


@router.post("/mark-all-read", response_model=MarkAllReadResponse)
async def mark_all_read(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> MarkAllReadResponse:
    return await svc.mark_all_read(session, user.id)


@router.delete(
    "/{notification_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_class=Response,
)
async def delete_notification(
    notification_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> Response:
    if not await svc.delete(session, user.id, notification_id):
        raise HTTPException(status_code=404, detail="notification not found")
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: apps/web/lib/api/notifications.ts
import type { paths } from "@ims/api-types"

import { apiFetch } from "./client"

export type NotificationRead =
  paths["/api/notifications"]["get"]["responses"]["200"]["content"]["application/json"]["items"][number]

export type NotificationListResponse =
  paths["/api/notifications"]["get"]["responses"]["200"]["content"]["application/json"]

export type UnreadCountResponse =
  paths["/api/notifications/unread-count"]["get"]["responses"]["200"]["content"]["application/json"]

export async function listNotifications(
  params: { unreadOnly?: boolean; limit?: number; offset?: number },
  accessToken: string | null,
): Promise<NotificationListResponse> {
  const q = new URLSearchParams()
  if (params.unreadOnly) q.set("unread_only", "true")
  if (params.limit !== undefined) q.set("limit", String(params.limit))
  if (params.offset !== undefined) q.set("offset", String(params.offset))
  const res = await apiFetch(`/notifications${q.size ? `?${q}` : ""}`, { accessToken })
  return (await res.json()) as NotificationListResponse
}

export async function getUnreadCount(accessToken: string | null): Promise<UnreadCountResponse> {
  const res = await apiFetch("/notifications/unread-count", { accessToken })
  return (await res.json()) as UnreadCountResponse
}

export async function markNotificationRead(
  id: string,
  accessToken: string | null,
): Promise<NotificationRead> {
  const res = await apiFetch(`/notifications/${id}/read`, { method: "PATCH", accessToken })
  return (await res.json()) as NotificationRead
}

export async function markAllNotificationsRead(
  accessToken: string | null,
): Promise<{ marked: number }> {
  const res = await apiFetch("/notifications/mark-all-read", { method: "POST", accessToken })
  return (await res.json()) as { marked: number }
}

export async function deleteNotification(
  id: string,
  accessToken: string | null,
): Promise<void> {
  await apiFetch(`/notifications/${id}`, { method: "DELETE", accessToken })
}
